### PR TITLE
Added linting for `spark.sql(...)` calls

### LINKED ISSUE #2551: [BUG]: Only table-migration linting occurs for queries ran via spark.sql("some query...")
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The query passed for execution to `spark.sql(...)` is only analyzed for table names.

### Expected Behavior

The query should be fully linted, using all SQL linters.

### Steps To Reproduce

_No response_

### Cloud

AWS

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
When linting queries in `spark.sql(...)` , current SparkSqlPyLinter linter only checks for table migration but does not use other SQL linters (such as DirectFsAccess linters).
This PR fixes the issue.

### Linked issues
Resolves #2551

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/base.py [hunk 1]
```diff
@@ -176,12 +176,22 @@ def lint(self, code: str) -> Iterable[Advice]:
...
 
 
-class Fixer:
+class Fixer(ABC):
     @abstractmethod
-    def name(self) -> str: ...
+    def can_fix(self, advice_code: str) -> bool: ...
+
+    @abstractmethod
+    def apply(self, advice_code: str, source_code: str) -> str: ...
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/base.py:184]: this is Fixer interface that has nothing to do with this PR. revert this change.

File: src/databricks/labs/ucx/source_code/linters/pyspark.py [hunk 2]
```diff
@@ -353,3 +319,59 @@ def _find_matcher(self, node: NodeNG):
...
+
+    def can_fix(self, advice_code: str) -> bool:
+        return advice_code.startswith(self._PREFIX)
+
+    def lint_tree(self, tree: Tree) -> Iterable[Advice]:
+        for call_node, query in self._visit_call_nodes(tree):
+            for value in InferredValue.infer_from_node(query):
+                if not value.is_inferred():
+                    yield Advisory.from_node(
+                        code=f"{self._PREFIX}cannot-autofix-table-reference",
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/pyspark.py:340]:
```suggestion
                        code=f"cannot-autofix-table-reference",
```
it's still the same problem code.

File: src/databricks/labs/ucx/source_code/linters/pyspark.py [hunk 3]
```diff
@@ -353,3 +319,59 @@ def _find_matcher(self, node: NodeNG):
...
+                        message=f"Can't migrate table_name argument in '{query.as_string()}' because its value cannot be computed",
+                        node=call_node,
+                    )
+                    continue
+                for advice in self._sql_linter.lint(value.as_string()):
+                    advice = advice.replace(code=f"{self._PREFIX}{advice.code}")
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/pyspark.py:346]:
```suggestion
```
i don't think it's required.

**Review by nfx — CHANGES_REQUESTED** [review]
revert changes that are out of scope of this PR

File: src/databricks/labs/ucx/source_code/base.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/base.py:184]: It's in scope because SparkSqlPyLinter is both a linter and a fixer. It needs to route `apply` to the correct SQL fixer.

**Comment by ericvergnaud** [comment]
The situation is that we have a top-level linter (SparkSqlPyLinter) that relies on a collection of SQL linters.
The issue is that SparkSqlPyLinter is also a Fixer, and relies on a collection of SQL Fixers.
When `apply` is invoked, it needs to extract the SQL, and route the advice to the correct SQL Fixer.
My proposal is to use a prefix in the advice, that SparkSqlPyLinter manages, but I'm happy to consider other ideas.
We can also have a specific SparkSqlPyLinter per SQLLinter. which removes the need for routing.

File: src/databricks/labs/ucx/source_code/base.py [hunk 4]
```diff
@@ -176,7 +176,9 @@ def lint(self, code: str) -> Iterable[Advice]:
     def lint_tree(self, tree: Tree) -> Iterable[Advice]: ...
 
 
-class Fixer:
+class Fixer(ABC):
+
+    @property
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/base.py:181]: fixes an inconsistency

**Comment by ericvergnaud** [comment]
> The situation is that we have a top-level linter (SparkSqlPyLinter) that relies on a collection of SQL linters. The issue is that SparkSqlPyLinter is also a Fixer, and relies on a collection of SQL Fixers. When `apply` is invoked, it needs to extract the SQL, and route the advice to the correct SQL Fixer. My proposal is to use a prefix in the advice, that SparkSqlPyLinter manages, but I'm happy to consider other ideas. We can also have a specific SparkSqlPyLinter per SQLLinter. which removes the need for internal routing.

@nfx I've refactored to the latter: 1 SparkSqlPyLinter per SQLLinter

**Review by nfx — APPROVED** [review]
lgtm
